backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/bll/VmHandler.java

```diff
@@ -1309,6 +1315,29 @@ public void updateCpuAndNumaPinning(VM vm, Guid vdsId) {
         }
     }
 
+    public void updateEmptyCpuPinningWithNuma(VM vm, Guid vdsId) {
+        String cpuPinning = CpuPinningHelper.getVmPinning(vm);
+        if (StringUtils.isNotEmpty(cpuPinning)) {
+            return;
+        }
+        List<VmNumaNode> vmNumaNodes = vmInfoBuildUtils.getVmNumaNodes(vm);
+        List<VdsNumaNode> vdsNumaNodes = vdsNumaNodeDao.getAllVdsNumaNodeByVdsId(vdsId);
+        boolean numaEnabled = vmInfoBuildUtils.isNumaEnabled(vdsNumaNodes, vmNumaNodes, vm);
+        if (!numaEnabled) {
+            return;
+        }
+        Map <String, Object> pinningMap = NumaSettingFactory.buildCpuPinningWithNumaSetting(
+                vmNumaNodes,
+                vdsNumaNodes);
+        // The policy not saved to the DB.
+        vm.setCpuPinningPolicy(CpuPinningPolicy.MANUAL);
```

Comment:
do I interpret your comment above as if you set the policy here without storing it in the DB later on so we could tell VDSM that the policy is MANUAL?
if so, wouldn't it make more sense to move that logic to LibvirtVmXmlBuilder and to say there that when we write a VM that is set with policy NONE and yet its "current" CPU pinning is not empty, we write MANUAL into the metadata? if we do that just for VDSM then that would be a better place to do that..

Reply:
Not only, we also need it for the allocation VdsCpuUnit. The logic is based on the policy we are with. Here I set it and it will be `MANUAL` for here forward. But, it won't be saved to the `VmStatic`. If we go with your approach we will need to check it also in the physical CPU allocation. What is better in your opinion?

Reply:
I see, so maybe a new entry in CpuPinningPolicy should be added then - something that would be translated to MANUAL when passed to VDSM and respected by other part in the engine. I'm worried about us forgetting this case and going to the static pinning when the policy is MANUAL later on

Reply:
Please let's try to avoid adding a new entry to CpuPinningPolicy if not absolutely necessary as it requires changes in REST API, Webadmin UI, ui-extensions...

Reply:
> Please let's try to avoid adding a new entry to CpuPinningPolicy if not absolutely necessary as it requires changes in REST API, Webadmin UI, ui-extensions...

I agree.
The way it's done in the code is solving the allocation+domxml VDSM problem without making further changes. Nobody "noticing" it except if you look on the code, like it was done before when not having CPU pinning with NUMA pinning.

Reply:
When in doubt add more comments.

Reply:
The dynamic pinning will be saved.
I see no reason to add a new policy, also we do need to set it here, otherwise it's possible to do without and we just need to check in the CPU allocation this situation also, when writing to policy to the metadata. But, I see it harmless as it's not saved anyway - it's runtime only, for the above.

Reply:
> Please let's try to avoid adding a new entry to CpuPinningPolicy if not absolutely necessary as it requires changes in REST API, Webadmin UI, ui-extensions...

not necessarily - see the vm-backup states for instance, we have bunch of them but only a subset is exposed to clients
we shouldn't be afraid from adding that, but we need to do it wisely if we go this way